src/librustc_lexer/src/lib.rs

```diff
@@ -476,99 +268,123 @@ impl Cursor<'_> {
             }
-            'e' | 'E' => {
+            c if self.is_exponent_start(c) => {
                 self.bump();
-                let empty_exponent = self.float_exponent().is_err();
+                let empty_exponent = !self.eat_float_exponent();
                 Float { base, empty_exponent }
             }
             _ => Int { base, empty_int: false },
         }
     }
 
+    /// Checks whether upcoming symbols are fraction part of the number.
+    fn is_fraction_start(&self, c: char) -> bool {
+        // Don't be greedy if this is actually an
+        // integer literal followed by field/method access or a range pattern
+        // (`0..2` and `12.foo()`)
+        c == '.' && self.second() != '.' && !is_id_start(self.second())
+    }
+
+    /// Checks whether upcoming symbols are exponent of the number.
+    fn is_exponent_start(&self, c: char) -> bool {
+        match c {
+            'e' | 'E' => true,
+            _ => false,
+        }
+    }
+
     fn lifetime_or_char(&mut self) -> TokenKind {
         debug_assert!(self.prev() == '\'');
-        let mut starts_with_number = false;
-
-        // Check if the first symbol after '\'' is a valid identifier
-        // character or a number (not a digit followed by '\'').
-        if (is_id_start(self.nth_char(0))
-            || self.nth_char(0).is_digit(10) && {
-                starts_with_number = true;
-                true
-            })
-            && self.nth_char(1) != '\''
-        {
-            self.bump();
 
-            // Skip the identifier.
-            while is_id_continue(self.nth_char(0)) {
-                self.bump();
+        if self.is_char_literal() {
+            let terminated = self.single_quoted_string();
+            let suffix_start = self.len_consumed();
+            if terminated {
+                self.eat_literal_suffix();
             }
+            let kind = Char { terminated };
+            return Literal { kind, suffix_start };
+        }
 
-            return if self.nth_char(0) == '\'' {
-                self.bump();
-                let kind = Char { terminated: true };
-                Literal { kind, suffix_start: self.len_consumed() }
-            } else {
-                Lifetime { starts_with_number }
-            };
+        // Either a lifetime or a character literal with
+        // length greater than 1.
+
+        let starts_with_number = self.first().is_digit(10);
+
+        // Skip the literal contents.
+        // First symbol can be a number (which isn't a valid identifier start),
+        // so skip it without any checks.
+        self.bump();
+        self.eat_identifier_continuation();
+
+        // Check if after skipping literal contents we've met a closing
+        // single quote (which means that user attempted to create a
+        // string with single quotes).
+        if self.first() == '\'' {
+            self.bump();
+            let kind = Char { terminated: true };
```

Comment:
By the way, I'm not sure why we're consuming the literal suffix above, but do not consume here.
As a result, we have a different errors for non-single character single-quoted literals with suffixes depending on the first symbol:
[Playground 1](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=27a912f94979701a1578e897afca1f05) / [Playground 2](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e237273802cc05dad098e265a8fb2e05)

That's pretty esoteric, I know, but nevertheless it seems a bit inconsistent to me.

Reply:
(I'm not sure if that can even be called a bug since the code in example is completely invalid)